### PR TITLE
Created formal fortran namelist for itag.

### LINKED ISSUE #115: Make itag fortran namelist
Request from UFS MRweather APP development.  View[ Issue 65](https://github.com/ufs-community/ufs-mrweather-app/issues/65) on ufs-meweather-app repo.

### PR DESCRIPTION
**Pull request by kayeekayee**
This code changes have been test on Cheyenne and Hera.
It is bitwise for both EMC and DTC regression test.

Latest code has been tested on 2 locations:
/scratch2/BMC/det/KaYee/UPP/UFFDA/itag/  (Hera)
/glade/scratch/kayee/UPP/itag/UFFDA (cheyenne)

These are the files being changed:
/scratch2/BMC/det/KaYee/UPP/UFFDA/itag/UPP_itag
modified:   scripts/run_upp
modified:   sorc/ncep_post.fd/WRFPOST.f
modified:   ush/gfs_nceppost.sh

The EMC regression test scripts are also modified for this code on Hera:
/scratch2/BMC/det/KaYee/UPP/UFFDA/itag/EMC_RT_itag/jobs
run_post_fv3gfsv16_pe_test_HERA.sh
run_post_fv3gfsv16_HERA.sh
run_post_fv3gfs_pe_test_HERA.sh
run_post_fv3gfs_HERA.sh
run_post_rap_HERA.sh
run_post_rap_pe_test_HERA.sh
run_post_hrrr_HERA.sh
run_post_hrrr_pe_test_HERA.sh
run_post_nmmb_Grib2_HERA.sh
run_post_nmmb_Grib2_pe_test_HERA.sh
run_post_fv3r_HERA.sh
run_post_fv3r_pe_test_HERA.sh
run_post_fv3gefs_HERA.sh
run_post_fv3gefs_pe_test_HERA.sh
run_post_fv3hafs_HERA.sh
run_post_fv3hafs_pe_test_HERA.sh
run_post_3drtma_HERA.sh
run_post_3drtma_pe_test_HERA.sh

/scratch2/BMC/det/KaYee/UPP/UFFDA/itag/EMC_RT_itag/gfs/ush
gfs_nceppost.sh

### CODE REVIEW

**Comment by WenMeng-NOAA** [comment]
@kayeekayee Can you create a new issue regarding this PR?

**Comment by fossell** [comment]
> @kayeekayee Can you create a new issue regarding this PR?

There is an existing issue already. I'll link it to this PR.

**Comment by WenMeng-NOAA** [comment]
@kayeekayee and @fossell I did tests on WCOSS-Dell. The results look good. There will be  coordination with model developers for this itag update. I would expect three more weeks for final merge.

File: ush/global_nceppost.sh [hunk 1]
```diff
@@ -350,13 +350,15 @@ export DD=`echo $VDATE | cut -c7-8`
...
 
 cat > itag <<EOF
-$GFSOUT
-${MODEL_OUT_FORM}
-${GRIBVERSION}
-${YY}-${MM}-${DD}_${HH}:00:00
-GFS
-$FLXINP
-$D3DINP
+&read_itag
```
Comment (WenMeng-NOAA) [inline ush/global_nceppost.sh]: @kayeekayee This is a legacy script. No updates are needed. Please revert the changes.

**Comment by WenMeng-NOAA** [comment]
@kayeekayee Could you sync your branch with upstream/develop? Thanks!

File: sorc/ncep_post.fd/WRFPOST.f [hunk 2]
```diff
@@ -215,34 +218,42 @@ PROGRAM WRFPOST
...
-!      if(MODELNAME == 'NMM')then
+!KaYee: Read itag in Fortran Namelist format
+!Initialize the namelist variables
+       fileName='gfs.t00z.atmf006.nc'
+       IOFORM='netcdf'
+       grib='grib2'
+       DateStr='2020-02-04_00:00:00'
+       FULLMODELNAME='GFS'
+       fileNameFlux='gfs.t00z.sfcf006.nc'
+       fileNameD3D='postxconfig-NT.txt'
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/WRFPOST.f]: @kayeekayee These lines might not be needed.

File: sorc/ncep_post.fd/WRFPOST.f [hunk 3]
```diff
@@ -215,34 +218,42 @@ PROGRAM WRFPOST
...
+888    if (itag_ierr /= 0) then
+       print*,'Incorrect namelist variable(s) found in the itag file,stopping!'
+       stop
+       endif
+       if (me==0) print*,'fileName= ',fileName
+         if (me==0) print*,'IOFORM= ',IOFORM
+         !if (me==0) print*,'OUTFORM= ',grib
+!        if(index(grib,"grib") == 0) then
+!          grib='grib1'
+!        endif
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/WRFPOST.f]: @kayeekayee Please remove these comment lines.

File: sorc/ncep_post.fd/WRFPOST.f [hunk 4]
```diff
@@ -342,15 +347,15 @@ PROGRAM WRFPOST
...
-          if(MODELNAME == 'RAPR')then
-            read(5,*) (po(l),l=1,kpo)
+!KaYee          if(MODELNAME == 'RAPR')then
+!KaYee            read(5,*) (po(l),l=1,kpo)
 ! CRA READ VALID TIME UNITS
-            read(5,121) VTIMEUNITS
-            if(me == 0) then
-              print*,'VALID TIME UNITS = ', VTIMEUNITS
-            endif
+!KaYee            read(5,121) VTIMEUNITS
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/WRFPOST.f:340]: @kayeekayee I would like to keep reading vtimeunits as a option. Could add the change in line 176 as:
namelist/nampgb/kpo,po,kth,th,kpv,pv,fileNameAER,d3d_on,gocart_on,popascal &
                     ,hyb_sigp,rdaod,aqfcmaq_on,vtimeunits

File: sorc/ncep_post.fd/WRFPOST.f [hunk 5]
```diff
@@ -175,6 +175,9 @@ PROGRAM WRFPOST
       real,dimension(komax) :: po,th,pv
       namelist/nampgb/kpo,po,kth,th,kpv,pv,fileNameAER,d3d_on,gocart_on,popascal &
                      ,hyb_sigp,rdaod,aqfcmaq_on
+      integer      :: itag_ierr
+      namelist/read_itag/fileName,IOFORM,grib,DateStr,FULLMODELNAME,fileNameFlux &
```
Comment (fossell) [inline sorc/ncep_post.fd/WRFPOST.f]: @WenMeng-NOAA -Do you have a preference for what the namelist is called? Currently "read_itag" but could be something else to reflect its contents if you prefer.

**Comment by fossell** [comment]
@kayeekayee - Can you please change this PR to active by clicking "ready for review"? It's currently still a Draft PR.

File: sorc/ncep_post.fd/WRFPOST.f (hunk 5 above)
Comment (WenMeng-NOAA) [reply sorc/ncep_post.fd/WRFPOST.f]: @fossell and @kayeekayee  Could we name  "model_inputs" or other names which might make sense?

File: sorc/ncep_post.fd/WRFPOST.f (hunk 2 above)
Comment (kayeekayee) [reply sorc/ncep_post.fd/WRFPOST.f]: @WenMeng-NOAA I commented them out and we need to test it if it works. I encountered some problems at the beginning stage without this initialization.

File: sorc/ncep_post.fd/WRFPOST.f [hunk 6]
```diff
@@ -174,7 +174,10 @@ PROGRAM WRFPOST
       integer      :: kpo,kth,kpv
       real,dimension(komax) :: po,th,pv
       namelist/nampgb/kpo,po,kth,th,kpv,pv,fileNameAER,d3d_on,gocart_on,popascal &
-                     ,hyb_sigp,rdaod,aqfcmaq_on
+                     ,hyb_sigp,rdaod,aqfcmaq_on,vtimeunits
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/WRFPOST.f:178]: @kayeekayee Could you add a change log with your name after line 44?

File: sorc/ncep_post.fd/WRFPOST.f [hunk 7]
```diff
@@ -215,34 +219,39 @@ PROGRAM WRFPOST
...
+         if (me==0) print*,'IOFORM= ',IOFORM
+         !if (me==0) print*,'OUTFORM= ',grib
+         if (me==0) print*,'OUTFORM= ',grib
+         if (me==0) print*,'DateStr= ',DateStr
+         if (me==0) print*,'FULLMODELNAME= ',FULLMODELNAME
+         MODELNAME=FULLMODELNAME(1:4)
+         SUBMODELNAME=FULLMODELNAME(5:)
+       IF(len_trim(FULLMODELNAME)<5) THEN
+          SUBMODELNAME='NONE'
+       ENDIF
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/WRFPOST.f]: @kayeekayee I am wonder if you could replace "fullmodelname" with "modelname,submodelname" in line 179 so that modelname and submodelname can be directly read from namelist "model_inputs"?

Comment (WenMeng-NOAA) [reply sorc/ncep_post.fd/WRFPOST.f]: @kayeekayee Please let me know if you could modify for my proposal? Thanks!

File: ush/gfs_nceppost.sh [hunk 8]
```diff
@@ -311,13 +311,15 @@ export DD=`echo $VDATE | cut -c7-8`
...
-${YY}-${MM}-${DD}_${HH}:00:00
-GFS
-$FLXINP
-$D3DINP
+&model_inputs
+fileName='${GFSOUT}'
+IOFORM='${MODEL_OUT_FORM}'
+grib='${GRIBVERSION}'
+DateStr='${YY}-${MM}-${DD}_${HH}:00:00'
+FULLMODELNAME='GFS'
```
Comment (WenMeng-NOAA) [inline ush/gfs_nceppost.sh]: @kayeekayee Please change FULLMODELNAME into MODELNAME.

File: sorc/ncep_post.fd/WRFPOST.f [hunk 9]
```diff
@@ -215,34 +219,41 @@ PROGRAM WRFPOST
...
+       if (me==0) print*,'fileName= ',fileName
+         if (me==0) print*,'IOFORM= ',IOFORM
+         !if (me==0) print*,'OUTFORM= ',grib
+         if (me==0) print*,'OUTFORM= ',grib
+         if (me==0) print*,'DateStr= ',DateStr
+         if (me==0) print*,'MODELNAME= ',MODELNAME
+         if (me==0) print*,'SUBMODELNAME= ',SUBMODELNAME
+         FULLMODELNAME=MODELNAME//SUBMODELNAME
+!         MODELNAME=FULLMODELNAME(1:4)
+!         SUBMODELNAME=FULLMODELNAME(5:)
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/WRFPOST.f]: @kayeekayee With modelname and submodelname read from namelist, you may delete these lines.

File: sorc/ncep_post.fd/WRFPOST.f (hunk 7 above)
Comment (kayeekayee) [reply sorc/ncep_post.fd/WRFPOST.f]: Hi Wen,
I made the changes according to your request, however, I got one non-bitwise regression test result:
10/25 19:44:37Z -rtma test: your new post executable did not generate bit-identical WRFTWO.GrbF00 as the trunk
I'm looking into the problem. Just in case you want to take a look:
/scratch2/BMC/det/KaYee/UPP/UFFDA/itag/EMC_RT_itag/scripts

Comment (WenMeng-NOAA) [reply sorc/ncep_post.fd/WRFPOST.f]: @kayeekayee The rtma test needs reading both modelname and submodelname. You may add changes in itag as:
MODELNAME='RAPR'
SUBMODELNMAE='RTMA'

File: sorc/ncep_post.fd/WRFPOST.f (hunk 9 above)
Comment (kayeekayee) [reply sorc/ncep_post.fd/WRFPOST.f]: Can I remove line 246? I see line 258, 261, 264 are calling FULLMODELNAME.

Comment (WenMeng-NOAA) [reply sorc/ncep_post.fd/WRFPOST.f]: @kayeekayee I would think line 246-250 and 261 can be deleted. You might modify 258 and 264 properly since fullmodelname is not needed any more. Thanks!

File: sorc/ncep_post.fd/WRFPOST.f (hunk 7 above)
Comment (kayeekayee) [reply sorc/ncep_post.fd/WRFPOST.f]: Yes, the rtma test is passed as well. Thank you Wen.

File: sorc/ncep_post.fd/WRFPOST.f [hunk 10]
```diff
@@ -270,18 +265,12 @@ PROGRAM WRFPOST
...
         if (me==0) print*,'MODELNAME= ',MODELNAME,'grib=',grib
-!Chuang: If model is GFS, read in flux file name from unit5
         if(MODELNAME == 'GFS' .OR. MODELNAME == 'FV3R') then
-          read(5,111,end=117) fileNameFlux
           if (me == 0) print*,'first two file names in GFS or FV3= '  &
                                ,trim(fileName),trim(fileNameFlux)
- 117      continue
-
-          read(5,111,end=118) fileNameD3D
           if (me == 0) print*,'D3D names in GFS= ',trim(fileNameD3D)
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/WRFPOST.f]: @kayeekayee Please comment out this line since no fileNemeD3D is needed for GFS right now.

File: sorc/ncep_post.fd/WRFPOST.f [hunk 11]
```diff
@@ -215,42 +219,33 @@ PROGRAM WRFPOST
...
-        if (me==0) print*,'OUTFORM2= ',grib
-        read(5,112) DateStr
-        read(5,114) FULLMODELNAME
-        MODELNAME=FULLMODELNAME(1:4)
-        SUBMODELNAME=FULLMODELNAME(5:)
-      IF(len_trim(FULLMODELNAME)<5) THEN
-         SUBMODELNAME='NONE'
-      ENDIF
-!      if(MODELNAME == 'NMM')then
+!KaYee: Read itag in Fortran Namelist format
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/WRFPOST.f:222]: @kayeekayee Could you add a line of setting default submodule name? e.g.
SUBMODELNAME=''

Comment (WenMeng-NOAA) [reply sorc/ncep_post.fd/WRFPOST.f:222]: @kayeekayee Let's set default submodelname as 'NONE'. Thanks!

File: sorc/ncep_post.fd/WRFPOST.f (hunk 10 above)
Comment (kayeekayee) [reply sorc/ncep_post.fd/WRFPOST.f]: @WenMeng-NOAA I can delete that print statement, how about this one? 
call baopenr(iunitd3d,trim(fileNameD3D),iostatusD3D)
Is it ok to leave it like that?

Comment (WenMeng-NOAA) [reply sorc/ncep_post.fd/WRFPOST.f]: @kayeekayee Yes, you may delete this printout statment and leave "call baopenr" statement. Eventually I will clean up fileNameD3D in another PR. Thanks!

**Review by WenMeng-NOAA — APPROVED** [review]

**Comment by WenMeng-NOAA** [comment]
The UPP regression tests were successfully passed on WCOSS-Dell, WCOSS-Cray and Hera. This PR is ready for merging.

**Comment by WenMeng-NOAA** [comment]
@fossell and @hertneky Do you have more comments on this PR?

**Comment by kayeekayee** [comment]
> The UPP regression tests were successfully passed on WCOSS-Dell, WCOSS-Cray and Hera. This PR is ready for merging.

Thank you so much for your confirmation, Wen.

**Review by fossell — APPROVED** [review]
No additional comments from me.

File: scripts/run_upp [hunk 12]
```diff
@@ -323,13 +323,15 @@ fi
...
-${flxFileName}
-postxconfig-NT.txt
+&model_inputs
+fileName='${inFileName}'
+IOFORM='${inFormat}'
+grib='${outFormat}'
+DateStr='${YY}-${MM}-${DD}_${HH}:00:00'
+MODELNAME='${tag}'
+fileNameFlux='${flxFileName}'
+fileNameD3D='postxconfig-NT.txt'
```
Comment (hertneky) [inline scripts/run_upp]: @kayeekayee Is fileNameD3D in reference to D3D options? Is there a reason for using this name in our community run script? (Note that I am not familiar with D3D myself)

Comment (WenMeng-NOAA) [reply scripts/run_upp]: @hertneky The filenameD3D is a legacy model file name for GFS which is not used any more. Assigning UPP control file as fileNameD3D would be misleading. It would be better to assign it as a empty string.

Comment (WenMeng-NOAA) [reply scripts/run_upp]: Or no line for fileNameD3D in itag.

Comment (kayeekayee) [reply scripts/run_upp]: @hertneky Do you have any preference of the name for the postxconfig.txt? FileCntrltxt maybe?

Comment (hertneky) [reply scripts/run_upp]: Can it be called fileNameFlat, like it is in WRFPOST for consistency?

Comment (kayeekayee) [reply scripts/run_upp]: @WenMeng-NOAA @hertneky Code has been updated with fileNameFlat. It passed both EMC and DTC regression test. Please review.

File: ush/gfs_nceppost.sh [hunk 13]
```diff
@@ -311,13 +311,15 @@ export DD=`echo $VDATE | cut -c7-8`
...
-$FLXINP
-$D3DINP
+&model_inputs
+fileName='${GFSOUT}'
+IOFORM='${MODEL_OUT_FORM}'
+grib='${GRIBVERSION}'
+DateStr='${YY}-${MM}-${DD}_${HH}:00:00'
+MODELNAME='GFS'
+fileNameFlux='${FLXINP}'
+fileNameFlat='${D3DINP}'
```
Comment (WenMeng-NOAA) [inline ush/gfs_nceppost.sh]: @kayeekayee There is no need for the line fileNameFlat in gfs itag. Please remove line 321.

Comment (kayeekayee) [reply ush/gfs_nceppost.sh]: Done.

File: sorc/ncep_post.fd/WRFPOST.f [hunk 14]
```diff
@@ -380,13 +370,13 @@ PROGRAM WRFPOST
  116    continue
 !set control file name
         fileNameFlat='postxconfig-NT.txt'
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/WRFPOST.f]: @kayeekayee Please move line 371-372 to before line 308 for setting default fileNameFlat.

**Review by hertneky — APPROVED** [review]
I have no other comments

**Comment by WenMeng-NOAA** [comment]
The tests on Hera were passed. This PR is ready for merging.
